scRNA-seq/05-clustering_markers_scRNA.Rmd

```diff
@@ -0,0 +1,349 @@
+
+Assigning clusters is nice for visualization, but we would also like to be able to move toward a biological interpretation of the clusters and identifying the cell types in each cluster.
+To that end, we can identify marker genes that are differentially expressed among clusters. 
+
+It is worth noting here that the statistical calculations here are more than a bit circular: we identified clusters first based on gene expression, then we are using those same clusters to find differences in gene expression. 
+The result is that even if there were no *true* clusters, we would always find marker genes! 
+For a much more technical exploration of this circularity (and a method to correct for it), see a preprint by [Gao et al. (2020)](https://arxiv.org/abs/2012.02936).
+In light of this, it is better to think about marker gene identification as an aid in interpreting the clustering results (and possibly extending insights to new data sets), rather than results that should be interpreted on their own, and we should be extremely wary of justifying cluster assignments solely based on these results!
+With that caveat, let's proceed.
+
+To identify marker genes, we will use the `scran::findMarkers()` function, which will rank genes by their differential expression by calculating pairwise statistics among clusters.
+We have a few options for how to determine the gene rankings and marker gene list for each cluster. 
+At one end could include genes that are differentially expressed in *any* pairwise comparison against our focal cluster, or at the other we could only include genes that are differentially expressed in *all* comparisons with that cluster. 
+We could also do something in between, including genes that differentiate the focal cluster from *some* fraction of the other clusters.
+For now, we will use the `findMarkers()` function to rank the genes in each cluster by their combined scores against *all* other clusters, using the `pval.type` argument. 
+
+`findMarkers()` will return a list (technically a list-like object) of tables, one for each cell type, with statistics for each gene showing how well it differentiates that cell type against other types.
+
+
+```{r find_markers, live = TRUE}
+# use `findmarkers()` to calculate how well each gene  
+#  differentiates each cluster from *all* other clusters
+markers <- scran::findMarkers(hodgkins_sce, 
+                              groups = hodgkins_sce$nncluster, 
+                              pval.type = "all")
+```
+
+Next we can look at one of those tables.
+We will start with the first cluster, which we will select from the list using the R standard double bracket `[[1]]` notation.
+We also doing a bit of transformation here to pull the gene name into a column of its own.
+
+```{r marker_table}
+markers[[1]] %>%
+  as.data.frame() %>% # convert to a data frame
+  tibble::rownames_to_column("gene") %>% # make gene a column
+  dplyr::filter(FDR < 0.01) # keep only more significant markers
```

Comment:
Should we filter to the top n genes instead? My understanding is that the p-values (and then that extends to the adjusted p-values) are not valid outside of using them for ranking (per [OSCA](https://bioconductor.org/books/release/OSCA/marker-detection.html#p-value-invalidity)). We could `arrange(FDR)` and then `slice_max()` (that's what I think it is these days anyway!).  I know we're not _using this_ for anything but I think that would be closer to the analysis we want to model.

But perhaps `findMarkers()` is doing something fancier than `p.adjust()` under the hood.

Reply:
I see now that we're doing the arrange step when we save the files.

Reply:
Just going to talk to myself before returning this review... 

I guess we're going to hit these points really hard in the pathway analysis materials...

Reply:
No, you are absolutely right that this is bad! I think switching to `ntop` may be a better choice overall. I can emphasize in text that we are doing that b/c the p values are invalid. I am pretty sure that `findMarkers()` is not doing anything fancy. But my intution is rank order is probably still meaningful, at least in part? 

Fun fact: the data table is actually already sorted, but I wanted to act as if it might not be...

Reply:
`ntop` change sounds good. I don't want to give anyone the impression that they have the license to filter by a commonly used FDR threshold and just have that be it is all!